src/signatures/commitment_and_public_key_signature.rs

```diff
@@ -0,0 +1,349 @@
+    pub fn verify_challenge<'a, C>(
+        &self,
+        commitment: &'a HomomorphicCommitment<P>,
+        pubkey: &'a P,
+        challenge: &[u8],
+        factory: &C,
+    ) -> bool
+    where
+        for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Mul<&'b K, Output = P>,
+        for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Add<&'b P, Output = P>,
+        for<'b> &'b K: Mul<&'b K, Output = K>,
+        for<'b> &'b K: Add<&'b K, Output = K>,
+        C: HomomorphicCommitmentFactory<P = P>,
+    {
+        // The challenge must be a valid scalar
+        let e = match K::from_bytes(challenge) {
+            Ok(e) => e,
+            Err(_) => return false,
+        };
+
+        self.verify(commitment, pubkey, &e, factory)
+    }
+
+    /// Verify a signature on a commitment and group element statement using a given challenge (as a scalar)
+    pub fn verify<'a, C>(&self, commitment: &'a HomomorphicCommitment<P>, pubkey: &'a P, challenge: &K, factory: &C) -> bool
+    where
+        for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Mul<&'b K, Output = P>,
+        for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Add<&'b P, Output = P>,
+        for<'b> &'b K: Mul<&'b K, Output = K>,
+        for<'b> &'b K: Add<&'b K, Output = K>,
+        C: HomomorphicCommitmentFactory<P = P>,
+    {
+        // The challenge cannot be zero
+        if *challenge == K::default() {
+            return false
+        }
+
+        // Use a single weighted equation for verification to avoid unnecessary group operations
+        // For now, we use naive multiscalar multiplication, but offload the commitment computation
+        // This allows for the use of precomputation within the commitment itself, which is more efficient
+        // TODO: use curve library (Straus/Pippenger) multiscalar multiplication with partial precomputation
+        let mut rng = thread_rng();
+        let w = K::random(&mut rng); // must be random and not Fiat-Shamir!
+
+        // u_a*H + (u_x + w*u_y)*G == ephemeral_commitment + w*ephemeral_pubkey + e*commitment + (w*e)*pubkey
+        let verifier_lhs = factory.commit(&(&self.u_x + &(&w * &self.u_y)), &self.u_a).as_public_key().to_owned();
+        let verifier_rhs_unweighted = self.ephemeral_commitment.as_public_key() + (commitment * challenge).as_public_key();
+        let verifier_rhs_weighted = &self.ephemeral_pubkey * &w + pubkey * &(&w * challenge);
+
+        verifier_lhs == verifier_rhs_unweighted + verifier_rhs_weighted
+    }
+
+    /// Get the signature tuple `(ephemeral_commitment, ephemeral_pubkey, u_a, u_x, u_y)`
+    #[inline]
```

Comment:
These inlines are not necessary, and IMO should only be added if they are proven to add performance

Reply:
Removed for both signature types.